core/src/main/java/com/linecorp/armeria/common/stream/DeferredStreamMessage.java

```diff
@@ -81,10 +93,14 @@ protected void delegate(StreamMessage<T> delegate) {
             }).exceptionally(CompletionActions::log);
         }
 
-        @SuppressWarnings("unchecked")
-        final PendingSubscription<? super T> ps = pendingSubscription;
-        if (ps != null) {
-            subscribeToDelegate(delegate, ps);
+        for (;;) {
+            if (stateUpdater.compareAndSet(this, State.NONE, State.DELEGATED)) {
+                return;
+            }
+            if (stateUpdater.compareAndSet(this, State.SUBSCRIBED, State.COMPLETE)) {
```

Comment:
Could you add some comments here about why this seemingly infinite loop will never loop infinitely?